packages/aws-cdk/lib/api/aws-auth/sdk-provider.ts

```diff
@@ -126,49 +141,53 @@ export class SdkProvider {
    *
    * The `environment` parameter is resolved first (see `resolveEnvironment()`).
    */
-  public async forEnvironment(environment: cxapi.Environment, mode: Mode): Promise<ISDK> {
+  public async forEnvironment(environment: cxapi.Environment, mode: Mode, options?: CredentialsOptions): Promise<ISDK> {
     const env = await this.resolveEnvironment(environment);
-    const creds = await this.obtainCredentials(env.account, mode);
-    return new SDK(creds, env.region, this.sdkOptions);
-  }
-
-  /**
-   * Return an SDK which uses assumed role credentials
-   *
-   * The base credentials used to retrieve the assumed role credentials will be the
-   * same credentials returned by obtainCredentials if an environment and mode is passed,
-   * otherwise it will be the current credentials.
-   *
-   */
-  public async withAssumedRole(roleArn: string, externalId: string | undefined, environment: cxapi.Environment | undefined, mode: Mode | undefined) {
-    debug(`Assuming role '${roleArn}'.`);
+    const baseCreds = await this.obtainBaseCredentials(env.account, mode);
 
-    let region: string;
-    let masterCredentials: AWS.Credentials;
+    // At this point, we need at least SOME credentials
+    if (baseCreds.source === 'none') { throw new Error(fmtObtainCredentialsError(env.account, baseCreds)); }
 
-    if (environment !== undefined && mode !== undefined) {
-      const env = await this.resolveEnvironment(environment);
-      masterCredentials = await this.obtainCredentials(env.account, mode);
-      region = env.region;
-    } else {
-      region = this.defaultRegion;
-      masterCredentials = await this.defaultCredentials();
+    // Simple case is if we don't need to "assumeRole" here. If so, we must now have credentials for the right
+    // account.
+    if (options?.assumeRoleArn === undefined) {
+      if (baseCreds.source === 'incorrectDefault') { throw new Error(fmtObtainCredentialsError(env.account, baseCreds)); }
```

Comment:
curious: why should credentials from a plugin when there is no role to assume an error?

Reply:
That's not what this says though. This says if the default credentials are for the wrong account.